### PR TITLE
[libcerf] Update and fix

### PR DESCRIPTION
**Pull request by dg0yt**
Noticed in #40053 https://dev.azure.com/vcpkg/public/_build/results?buildId=105968&view=results.

### CODE REVIEW

File: ports/libcerf/begin-end-decls.diff [hunk 1]
```diff
@@ -0,0 +1,44 @@
+diff --git a/lib/cerf.h b/lib/cerf.h
+index 28576ab..c9bc6dd 100644
+--- a/lib/cerf.h
++++ b/lib/cerf.h
+@@ -55,25 +55,23 @@
+      typedef double _Complex _cerf_cmplx;
+ #  endif
+ 
+-#  undef __BEGIN_DECLS
+-#  undef __END_DECLS
```
Comment (dg0yt) [inline ports/libcerf/begin-end-decls.diff:10]: These `__` macros are used by gcc headers. Changing the compiler headers' behavior is neither desired nor feasible. The patch fixes an usual linking error on linux:
~~~
: && /usr/lib/ccache/c++ -fPIC -g  test/CMakeFiles/cerftest.dir/cerftest.c.o -o test/cerftest  lib/libcerfcpp.a  -lm && :
/usr/bin/ld: test/CMakeFiles/cerftest.dir/cerftest.c.o: in function `relerr(double, double)':
/vcpkg/buildtrees/libcerf/src/v2.4-4976f404d7/test/testtool.h:51: undefined reference to `__assert_fail(char const*, char const*, unsigned int, char const*)'
collect2: error: ld returned 1 exit status
~~~

File: ports/libcerf/cxx-flags.diff [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+     set(MS_NOWARN "/wd4018 /wd4068 /wd4101 /wd4244 /wd4267 /wd4305 /wd4715 /wd4996")
+     set(MS_D "-D_CRT_SECURE_NO_WARNINGS -D_SILENCE_TR1_NAMESPACE_DEPRECATION_WARNING")
+-    set(CMAKE_CXX_FLAGS_RELEASE "${CMAKE_CXX_FLAGS} ${MS_NOWARN} ${MS_D}")
++    set(CMAKE_CXX_FLAGS_RELEASE "${CMAKE_CXX_FLAGS_RELEASE} ${MS_NOWARN} ${MS_D}")
+     set(CTEST_CONFIGURATION_TYPE "${JOB_BUILD_CONFIGURATION}")
+     if(BUILD_SHARED_LIBS)
+         set(CMAKE_WINDOWS_EXPORT_ALL_SYMBOLS ON)
+-    else()
+-        # required for post-build validation under vcpkg:
+-        set(CMAKE_CXX_FLAGS_DEBUG "${CMAKE_CXX_FLAGS_DEBUG} /MTd") # multithreaded, debug
```
Comment (dg0yt) [inline ports/libcerf/cxx-flags.diff:19]: NB: The last lines were from a wrong patch, upstreamed from vcpkg.
CMake already applies per-config flags in addition to standard flags. The original problem was the overwriting of the per-config flags.

File: ports/libcerf/fix-source.diff [hunk 3]
```diff
@@ -0,0 +1,26 @@
+diff --git a/lib/width.c b/lib/width.c
+index 37c3afa..b4b6bda 100644
+--- a/lib/width.c
++++ b/lib/width.c
+@@ -47,7 +47,7 @@ using std::isnan;
+  * Olivero & Longbothum [1977]
+  * Journal of Quantitative Spectroscopy and Radiative Transfer. 17:233
+  */
+-double hwhm0(double sigma, double gamma)
++static double hwhm0(double sigma, double gamma)
```
Comment (dg0yt) [inline ports/libcerf/fix-source.diff:10]: `static` is in upstream's current `HEAD`.

**Review by FrankXie05 — APPROVED** [review]

**Comment by dg0yt** [comment]
FTR upstream is notified of the patches but won't react soon enough. IMO there is no reason to stop this PR: all changes are necessary at least for vcpkg, and it is no worse than before.

**Review by data-queue — APPROVED** [review]

**Comment by data-queue** [comment]
LGTM, however I want to give upstream some time to respond with any feedback.

**Comment by dg0yt** [comment]
They are on holidays. I can comment even more on all patches here.

**Pull request merged**